### PR TITLE
[entropy] allow for entropy source sharing

### PR DESCRIPTION
**Pull request by KaSroka**
This pr addresses #3798 

It consists of two commits, one lowering memory requirement for nRF52811 target and the second one with entropy sharing.

Work done:
* Random number generation and usage was cleaned.
* Module for cryptographic and software random number generator was created.
* Module for entropy source management was created.
* Platform is now responsible for providing entropy source.
* MbedTls context for entropy source and cryptographic random number generator was exposed to the application.
* Single entropy and cryptographic random generator context across whole OpenThread (and possibly application) is used.
* Entropy source can be pooled less often possibly lowering power usage.
* Easier integration with RTOS environment thanks to MbedTls built-in mutex support (can be enabled using MBEDTLS_THREADING_C define).

### CODE REVIEW

File: src/core/common/entropy.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+#endif // OPENTHREAD_RADIO
+
+namespace ot {
+namespace Entropy {
+
+static bool sInitialized;
+
+#ifndef OPENTHREAD_RADIO
+static mbedtls_entropy_context sEntropy;
```
Comment (jwhui) [inline src/core/common/entropy.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: src/core/common/entropy.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+#include <mbedtls/entropy.h>
+#include <mbedtls/entropy_poll.h>
+
+#endif // OPENTHREAD_RADIO
+
+namespace ot {
+namespace Entropy {
+
+static bool sInitialized;
```
Comment (jwhui) [inline src/core/common/entropy.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: src/core/common/random_crypto.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <mbedtls/entropy.h>
+
+#include "debug.hpp"
+
+namespace ot {
+namespace Random {
+namespace Crypto {
+
+static mbedtls_ctr_drbg_context sCtrDrbg;
```
Comment (jwhui) [inline src/core/common/random_crypto.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: src/core/common/random_crypto.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+static bool                     sInitialized;
```
Comment (jwhui) [inline src/core/common/random_crypto.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: include/openthread/random.h [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+ *
+ */
+
+/**
+ * This function generates and returns a random `uint32_t` value.
+ *
+ * @returns    A random `uint32_t` value.
+ *
+ */
+uint32_t otRandomSoftwareGetUint32(void);
```
Comment (jwhui) [inline include/openthread/random.h]: I understand you are trying to differentiate from a cryptographic PRNG. However, I think it's strange to use `Software` as the differentiator. I do not currently have a good suggestion other than to have `otRandomGet/Fill` as the prefix for non-cryptographic versions and `otRandomCryptoGet/Fill*` as the the prefix for cryptographic versions.

Thoughts?

**Comment by KaSroka** [comment]
@jwhui The idea was to have single entropy and ctr_drbg context even if we have multiple OT instances - so it is shared between all OT instances. Entropy and Random::Crypto module right now doesn't support multiple initialization/deinitialization but it can be easily added in the future (switch from bool flag initialized/not initialized to counter so we can initialize on first init and deinitialize on last deinit)

File: src/core/common/random_software.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#include "random.hpp"
+
+#include "debug.hpp"
+#include "entropy.hpp"
+
+namespace ot {
+namespace Random {
+namespace Software {
+
+static uint32_t mState;
```
Comment (jwhui) [inline src/core/common/random_software.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: src/core/common/random_software.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+static bool     sSeeded;
```
Comment (jwhui) [inline src/core/common/random_software.cpp]: Can/should this be included in [`ot::Instance`](https://github.com/openthread/openthread/blob/master/src/core/common/instance.hpp#L100)?

File: include/openthread/random.h (hunk 5 above)
Comment (KaSroka) [reply include/openthread/random.h]: I was having same trouble - the other idea I had was `Crypto` and `Weak`

Comment (KaSroka) [reply include/openthread/random.h]: I think we should put some keyword in software random generator functions so it makes a clear distinguishment between cryptographic and non-cryptographic functions. `Crypto` and `NonCrypto` is another option.

Comment (jwhui) [reply include/openthread/random.h]: `NonCrypto` works for me - I prefer that to `Software`.

**Comment by jwhui** [comment]
@KaSroka, thanks for the clarification. I think it makes sense to maintain a single `mbedtls_entropy_context` and `mbedtls_ctr_drbg_context` shared between all instances. Can we add the counter as part of this PR?

What about the MLCG? If we want to keep that static, how about a check on `sSeeded` to avoid re-seeding?

**Comment by jwhui** [comment]
@KaSroka, BTW, thanks for this contribution! This is a great enhancement.

**Comment by KaSroka** [comment]
> I think it makes sense to maintain a single `mbedtls_entropy_context` and `mbedtls_ctr_drbg_context` shared between all instances. Can we add the counter as part of this PR?

Sure, it should be quick to add.
> What about the MLCG? If we want to keep that static, how about a check on `sSeeded` to avoid re-seeding?

It depends if re-seeding is a bad thing or not - that's a thing to consider.

File: src/core/common/random_software.cpp (hunk 6 above)
Comment (jwhui) [inline src/core/common/random_software.cpp]: nit: rename to `sState`

**Comment by KaSroka** [comment]
Any idea what should I fix in `vcxproj` files to make AppVeyor happy?

**Comment by jwhui** [comment]
@KaSroka, Windows build uses `OPENTHREAD_ENABLE_MULTIPLE_INSTANCES`. It appears that `crypto/mbedtls.hpp` is not included, which is needed for `ot::Crypto::MbedTls`.

**Comment by KaSroka** [comment]
AppVeyor should be green now. I've also introduced all the current review changes.

File: examples/drivers/windows/otLwf/thread.c [hunk 8]
```diff
@@ -424,13 +424,7 @@ uint16_t otPlatMessagePoolNumFreeBuffers(_In_ otInstance *otCtx)
...
 #endif
 
-uint32_t otPlatRandomGet()
-{
-    LARGE_INTEGER Counter = KeQueryPerformanceCounter(NULL);
-    return (uint32_t)RtlRandomEx(&Counter.LowPart);
-}
-
-otError otPlatRandomGetTrue(uint8_t *aOutput, uint16_t aOutputLength)
+otError utilsEntropyGet(uint8_t *aOutput, uint16_t aOutputLength)
```
Comment (LuDuda) [inline examples/drivers/windows/otLwf/thread.c]: Leftover from initial implementation: utilsEntropyGet -> otPlatEntropyGet

File: tests/fuzz/fuzzer_platform.c [hunk 9]
```diff
@@ -339,33 +338,11 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
...
-        mlcg++;
-    }
-
-    sRandomState = mlcg;
-
-    return mlcg;
-}
-
-otError otPlatRandomGetTrue(uint8_t *aOutput, uint16_t aOutputLength)
+otError utilsEntropyGet(uint8_t *aOutput, uint16_t aOutputLength)
```
Comment (LuDuda) [inline tests/fuzz/fuzzer_platform.c]: Leftover from initial implementation: utilsEntropyGet -> otPlatEntropyGet

**Comment by KaSroka** [comment]
@LuDuda Nice catch, I think that we are missing fuzz platform tests.

**Comment by KaSroka** [comment]
Rebased

File: src/core/common/instance.hpp [hunk 10]
```diff
@@ -311,10 +313,14 @@ class Instance : public otInstance
     Notifier mNotifier;
     Settings mSettings;
 
-#if !OPENTHREAD_ENABLE_MULTIPLE_INSTANCES
-#if OPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS
     Crypto::MbedTls mMbedTls;
```
Comment (abtink) [inline src/core/common/instance.hpp]: With this we will have `mMbedTls` even when `OPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS` is not enabled. Is this intentional?

File: src/core/common/entropy.hpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * This function generates and returns a 32 bit entropy.
+ *
+ * @param aVal[out]  A generated 32 bit entropy value.
+ *
+ * @retval OT_ERROR_NONE    Successfully generated 32 bit entropy.
+ *
+ */
+otError GetUint32(uint32_t *aVal);
```
Comment (abtink) [inline src/core/common/entropy.hpp]: How about `uint32_t & aVal`?

File: src/core/common/instance.hpp (hunk 10 above)
Comment (KaSroka) [reply src/core/common/instance.hpp]: We still need `otError MbedTls::MapError(int rval)` even if external mbedTLS is used. `mbedtls.cpp` includes proper `#if` statements to exclude what's not needed.

File: src/core/common/entropy.hpp (hunk 11 above)
Comment (KaSroka) [reply src/core/common/entropy.hpp]: :+1:

**Comment by KaSroka** [comment]
Review changes applied

File: src/core/common/entropy.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <mbedtls/entropy.h>
+#include <mbedtls/entropy_poll.h>
+
+#endif // OPENTHREAD_RADIO
+
+namespace ot {
+namespace Entropy {
+
+static uint32_t sInitCnt;
```
Comment (bukepo) [inline src/core/common/entropy.cpp]: initialize as 0?

File: src/core/common/entropy.cpp [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+static mbedtls_entropy_context sEntropy;
+
+static int HandleMbedtlsEntropyPoll(void *aData, unsigned char *aOutput, size_t aInLen, size_t *aOutLen)
+{
+    OT_UNUSED_VARIABLE(aData);
+
+    otError error;
+    int     rval = 0;
+
+    error = otPlatEntropyGet((uint8_t *)aOutput, (uint16_t)aInLen);
```
Comment (bukepo) [inline src/core/common/entropy.cpp]: static_cast?

File: src/core/common/entropy.cpp (hunk 12 above)
Comment (KaSroka) [reply src/core/common/entropy.cpp]: :+1:

File: examples/platforms/cc1352/entropy.c [hunk 14]
```diff
@@ -122,9 +118,9 @@ static int TRNGPoll(unsigned char *aOutput, size_t aLen)
 }
 
 /**
- * Function documented in platform/random.h
+ * Function documented in platform/entropy.h.h
```
[the remark this answers is not in the document]
Comment (KaSroka) [reply examples/platforms/cc1352/entropy.c]: :+1:

File: src/core/common/entropy.cpp (hunk 13 above)
Comment (KaSroka) [reply src/core/common/entropy.cpp]: :+1:

**Comment by KaSroka** [comment]
Review changes applied

**Pull request merged**